File: 2020/25/Haskell/src/Naive.hs
module Naive where

modulus :: Int
modulus = 20201227

transformBy :: ((Int, Int) -> Bool) -> Int -> Int -> Int -> (Int, Int)
transformBy f i x x0 | x == (-1) = transformBy f 1 x0 x0
                     | f (x, i)  = (x, i)
                     | otherwise = transformBy f (i + 1) x' x0
  where x' = (x * x0) `mod` modulus

solve :: (Int, Int) -> Int
solve (k1, k2) = r1
 where
  l1 = snd . transformBy ((== k1) . fst) (-1) (-1) $ 7
  r1 = fst . transformBy ((== l1) . snd) (-1) (-1) $ k2

File: 2020/25/Rust/src/naive.rs
use crate::MODULUS;

fn transform_by<F>(predicate: F, subject_number: &u128) -> (u128, u128)
where
    F: Fn((u128, u128)) -> bool,
{
    let mut key = *subject_number;
    let mut loop_size = 1;
    while !predicate((key, loop_size)) {
        key = (key * subject_number) % MODULUS;
        loop_size += 1;
    }
    (key, loop_size)
}

pub fn solve(keys: &(u128, u128)) -> u128 {
    let (key_1, key_2) = *keys;
    let loop_size = transform_by(|x| x.0 == key_1, &7).1;
    transform_by(|x| x.1 == loop_size, &key_2).0
}
